src/UI/Components/Diagram/helpers.ts

```diff
@@ -284,45 +286,50 @@ export const shapesDataTransform = (
 
   for (const [key, instancesToEmbed] of Object.entries(groupedEmbedded)) {
     if (instance.attributes) {
-      if (instancesToEmbed.length > 1) {
-        const updated: InstanceAttributeModel[] = [];
-        instancesToEmbed.forEach((instance) => {
+      const updated: InstanceAttributeModel[] = [];
+      let shouldEmbeddedBeSingular = false;
+
+      //iterate through instancesToEmbed to recursively joint potential nested embedded entities into correct objects in correct state
+      instancesToEmbed.forEach((instance) => {
+        const embeddedModel = serviceModel.embedded_entities.find(
+          (entity) => entity.name === instancesToEmbed[0].service_entity,
+        );
+        if (embeddedModel) {
           const updatedInstance = shapesDataTransform(
             notMatchingInstances,
             instance,
-            true,
+            embeddedModel,
+            !!embeddedModel,
           );
 
-          areEmbeddedEdited =
-            !areEmbeddedEdited &&
-            !instance.action &&
-            updatedInstance.action !== null;
-
-          if (updatedInstance.action !== "delete") {
-            updated.push(updatedInstance.attributes as InstanceAttributeModel);
+          if (!areEmbeddedEdited) {
+            areEmbeddedEdited =
+              !instance.action && updatedInstance.action !== null;
           }
-        });
 
-        instance.attributes[key] = updated;
-      } else {
-        const data = shapesDataTransform(
-          notMatchingInstances,
-          instancesToEmbed[0],
-          true,
-        );
-
-        areEmbeddedEdited = instance.action === null && data.action !== null;
+          shouldEmbeddedBeSingular =
```

Comment:
I know performance is not much a concern, but allocating a variable like this isn't ideal. It would be more efficient to have a method to which you pass as argument the model and return whether it's singular or not. When I read the flow of code here, it makes me wonder why it has to be done between the two if cases too.

```
const isSingularRelation: boolean = (model: ModelType ) => { 
     !!model.upper_limit && model.upper_limit === 1;
}
```
I don't know by heart which name the Type is supposed to have for the model. But you get my drift. You can then simply pass the method in line 323. And also reuse the method wherever needed. If I remember well, it's a logic that appears now and then, to check if it's one or more relations.